src/NHibernate/Proxy/NHibernateProxyBuilder.cs

```diff
@@ -393,9 +391,28 @@ private static void ImplementGetLazyInitializer(TypeBuilder typeBuilder, MethodI
 			IL.Emit(OpCodes.Ldnull);
 			IL.Emit(OpCodes.Bne_Un, skipBaseCall);
 
-			IL.Emit(OpCodes.Ldarg_0);
-			EmitCallMethod(IL, OpCodes.Call, method);
-			IL.Emit(OpCodes.Ret);
+			if (method.IsAbstract)
+			{
+				if (!method.ReturnType.IsValueType)
+				{
+					IL.Emit(OpCodes.Ldnull);
+				}
+				else if (method.ReturnType != typeof(void))
+				{
+					var local = IL.DeclareLocal(method.ReturnType);
+					IL.Emit(OpCodes.Ldloca, local);
+					IL.Emit(OpCodes.Initobj, method.ReturnType);
+					IL.Emit(OpCodes.Ldloc, local);
+				}
+
+				IL.Emit(OpCodes.Ret);
+			}
+			else
```

Comment:
With the current tests this branch is not needed at all. If the `if` condition replace with `if(true)` all tests still pass.

I'm considering to make a breaking change and throw exception if `__lazyInitializer` is null.